### ISSUE TITLE
Concurrent map read and map write when multiple reads

### ISSUE THREAD

**Issue by philwinder** [weaveworks/flux#393]
I can consistently cause a crash by running:
`seq 1000 | parallel -j 10 'echo {} && fluxctl release --service="default/helloworld" --update-image="quay.io/weaveworks/helloworld:master-a000001" —dry-run'`

And waiting some time.

The original crash is being generated by the "/cznic/ql" driver:
```
github.com/weaveworks/flux/vendor/github.com/cznic/ql.newdriverRows.func1(0xc420ab8160)
	/Users/phil-mac/source/gocode/src/github.com/weaveworks/flux/vendor/github.com/cznic/ql/driver.go:394 +0xa4 fp=0xc420639f78 sp=0xc420639e88
runtime.goexit()
	/usr/local/opt/go/libexec/src/runtime/asm_amd64.s:2086 +0x1 fp=0xc420639f80 sp=0xc420639f78
created by github.com/weaveworks/flux/vendor/github.com/cznic/ql.newdriverRows
	/Users/phil-mac/source/gocode/src/github.com/weaveworks/flux/vendor/github.com/cznic/ql/driver.go:402 +0xeb
```

And I think the concurrent access is occuring in this goroutine:
```
goroutine 44 [semacquire]:
sync.runtime_Semacquire(0xc4200dda6c)
    /usr/local/opt/go/libexec/src/runtime/sema.go:47 +0x30
sync.(*RWMutex).Lock(0xc4200dda64)
    /usr/local/opt/go/libexec/src/sync/rwmutex.go:91 +0x98
github.com/weaveworks/flux/vendor/github.com/cznic/ql.(*DB).run1(0xc4200dda40, 0xc420c88240, 0x7f4affc09fc0, 0x1ff8568, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
    /Users/phil-mac/source/gocode/src/github.com/weaveworks/flux/vendor/github.com/cznic/ql/ql.go:1234 +0xfc1
github.com/weaveworks/flux/vendor/github.com/cznic/ql.(*DB).Execute(0xc4200dda40, 0xc420c88240, 0xc420447b60, 0x1, 0x1, 0x0, 0x0, 0x0, 0x0, 0x97, ...)
    /Users/phil-mac/source/gocode/src/github.com/weaveworks/flux/vendor/github.com/cznic/ql/ql.go:1192 +0x77b
github.com/weaveworks/flux/vendor/github.com/cznic/ql.(*driverConn).Begin(0xc420660780, 0x1f93101, 0xc42041e070, 0x0, 0x0)
    /Users/phil-mac/source/gocode/src/github.com/weaveworks/flux/vendor/github.com/cznic/ql/driver.go:249 +0x9b
database/sql.(*DB).begin(0xc4200864d0, 0x456501, 0xc420cb0500, 0xc4209b1300, 0x667880)
    /usr/local/opt/go/libexec/src/database/sql/sql.go:1170 +0x80
database/sql.(*DB).Begin(0xc4200864d0, 0xc4200868f0, 0x1486280, 0xc420bee600)
    /usr/local/opt/go/libexec/src/database/sql/sql.go:1153 +0x4c
github.com/weaveworks/flux/jobs.(*DatabaseStore).Transaction(0xc4203a31e0, 0xc4209b13a8, 0xc420086914, 0x2a)
    /Users/phil-mac/source/gocode/src/github.com/weaveworks/flux/jobs/db_store.go:448 +0x94
github.com/weaveworks/flux/jobs.(*DatabaseStore).UpdateJob(0xc4203a31e0, 0xc42068e740, 0x15, 0xc420140360, 0x24, 0xc420690550, 0x7, 0xc420690557, 0x7, 0x15d8ce0, ...)
    /Users/phil-mac/source/gocode/src/github.com/weaveworks/flux/jobs/db_store.go:386 +0x1f2
github.com/weaveworks/flux/jobs.(*instrumentedJobStore).UpdateJob(0xc4203a3560, 0xc42068e740, 0x15, 0xc420140360, 0x24, 0xc420690550, 0x7, 0xc420690557, 0x7, 0x15d8ce0, ...)
    /Users/phil-mac/source/gocode/src/github.com/weaveworks/flux/jobs/metrics.go:74 +0xf2
github.com/weaveworks/flux/release.(*Releaser).Handle.func2(0x16c5134, 0x1c, 0x0, 0x0, 0x0)
    /Users/phil-mac/source/gocode/src/github.com/weaveworks/flux/release/releaser.go:82 +0x169
github.com/weaveworks/flux/release.(*Releaser).Handle(0xc420260ac0, 0xc420662240, 0x7f4affc12aa0, 0xc4203a3560, 0x0, 0x0, 0x0, 0x0, 0x0)
    /Users/phil-mac/source/gocode/src/github.com/weaveworks/flux/release/releaser.go:85 +0x4a3
github.com/weaveworks/flux/jobs.(*Worker).Work(0xc420256b40)
    /Users/phil-mac/source/gocode/src/github.com/weaveworks/flux/jobs/worker.go:96 +0x10ba
created by main.main
    /Users/phil-mac/source/gocode/src/github.com/weaveworks/flux/cmd/fluxsvc/main.go:311 +0x2b1e
```

Which corresponds to the line:
```    tx, err := s.conn.(*sql.DB).Begin()
```
In the db_store file.

So basically, I think it’s possible for two threads to be accessing the Begin() function at the same time. Clicking through the methods, the eventual `Lock` call is quite late.

**Comment by cznic** [comment]
@philwinder Can you please provide a complete reproduction case? I tried the command in the OP, but it exits immediately w/o producing any output or crash with exit status 0. Thank you.

**Comment by sambooo** [comment]
The code in question (fluxsvc) no longer lives in this repo, nor does it use the `ql` driver.
